### PR TITLE
[runtime] QubitPlacer part 1

### PR DESCRIPTION
**Pull request by mpharrigan**
Splitting up #4678 

- The interface: `QubitPlacer`
- The fallback: `NaiveQubitPlacer`
- runtime hooks

some hacks around #4699

### CODE REVIEW

**Comment by mpharrigan** [comment]
factored out named topology stuff in #4701

**Comment by mpharrigan** [comment]
ptal @MichaelBroughton @tanujkhattar or anyone interested

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+import abc
+import dataclasses
+from typing import Dict, Any, Tuple, TYPE_CHECKING
+
+import numpy as np
+
+import cirq
+from cirq import _compat
+from cirq.devices.named_topologies import NamedTopology
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/qubit_placement.py:25]: Nit: module import.

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        rs: np.random.RandomState,
+    ) -> Tuple[cirq.FrozenCircuit, Dict[Any, cirq.Qid]]:
+        """Place a circuit with a given topology.
+
+        Args:
+            circuit: The circuit.
+            problem_topo: The topologies (i.e. connectivity) of the circuit.
+            shared_rt_info: A `cg.SharedRuntimeInfo` object that may contain additional info
+                to inform placement.
+            rs: A `RandomState` to enable pseudo-random placement strategies.
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/qubit_placement.py:47]: Nit: what does this return ?

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    """Don't do any qubit placement, use circuit qubits."""
+
+    def place_circuit(
+        self,
+        circuit: cirq.AbstractCircuit,
+        problem_topo: NamedTopology,
+        shared_rt_info: 'cg.SharedRuntimeInfo',
+        rs: np.random.RandomState,
+    ) -> Tuple[cirq.FrozenCircuit, Dict[Any, cirq.Qid]]:
+        return circuit.freeze(), {q: q for q in circuit.all_qubits()}
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/qubit_placement.py:62]: Question: Why aren't there any strict checks that the resulting circuit is valid under the given topology ? Do we want to allow users to have a Placer totally ignore the Topology ?

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        Args:
+            circuit: The circuit.
+            problem_topo: The topologies (i.e. connectivity) of the circuit.
+            shared_rt_info: A `cg.SharedRuntimeInfo` object that may contain additional info
+                to inform placement.
+            rs: A `RandomState` to enable pseudo-random placement strategies.
+        """
+
+
+@dataclasses.dataclass(frozen=True)
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/qubit_placement.py:51]: Question: Why is this a dataclass with no fields ?

File: cirq-google/cirq_google/workflow/quantum_runtime.py [hunk 5]
```diff
@@ -180,26 +212,38 @@ def execute(
...
 
     saver = _FilesystemSaver(base_data_dir=base_data_dir, run_id=run_id)
     saver.initialize(rt_config, shared_rt_info)
 
-    sampler = rt_config.processor.get_sampler()
     logger = _PrintLogger(n_total=len(executable_group))
     logger.initialize()
+
+    rs = np.random.RandomState(rt_config.random_seed)
+    exe: QuantumExecutable
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/quantum_runtime.py:231]: Does this declaration mess with the below loop ?

File: cirq-google/cirq_google/workflow/quantum_runtime.py [hunk 6]
```diff
@@ -180,26 +212,38 @@ def execute(
...
     executable_results = []
 
     saver = _FilesystemSaver(base_data_dir=base_data_dir, run_id=run_id)
     saver.initialize(rt_config, shared_rt_info)
 
-    sampler = rt_config.processor.get_sampler()
     logger = _PrintLogger(n_total=len(executable_group))
     logger.initialize()
+
+    rs = np.random.RandomState(rt_config.random_seed)
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/workflow/quantum_runtime.py:230]: Question: Should we give the seed to the `rt_config` or the randomstate generator itself ? It seems like down the line if we kept creating `randomState`s off of this seed we'd get a lot of duplicate random numbers.

**Review by MichaelBroughton — COMMENTED** [review]
Looks pretty good, just a few high level questions before we merge.

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 1 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:25]: done sorry

File: cirq-google/cirq_google/workflow/quantum_runtime.py (hunk 6 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/quantum_runtime.py:230]: two part answer

1. `rt_config` needs to be serializable and put in a database so we can groupby and agg the results. Seed is (much!) easier than the internal state of a random state
2. We don't create >1 randomstate based on this seed. We create one at the top of the function (here!) and it gets passed to any subroutines that require randomness.

File: cirq-google/cirq_google/workflow/quantum_runtime.py (hunk 5 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/quantum_runtime.py:231]: this is how you do type annotations for loop variables

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 2 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:47]: added

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 4 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:51]: serialization, str+repr, equality, and frozenness

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 3 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:62]: yes; this placer totally ignores the topology. Validation can be done by the device as part of the execution loop

**Comment by mpharrigan** [comment]
@MichaelBroughton ptal

I've also merged in #4377  #4701

File: cirq-core/cirq/devices/named_topologies.py [hunk 7]
```diff
@@ -137,6 +139,9 @@ def draw(self, ax=None, tilted: bool = True, **kwargs) -> Dict[Any, Tuple[int, i
     def _json_dict_(self) -> Dict[str, Any]:
         return dataclass_json_dict(self)
 
+    def __repr__(self) -> str:
+        return _compat.dataclass_repr(self)
+
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq-core/cirq/devices/named_topologies.py]: please see #4701 but dataclasses don't play nice with abstract base classes where I want different `__init__` methods for subclasses

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+if TYPE_CHECKING:
+    import cirq_google as cg
+
+
+class QubitPlacer(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def place_circuit(
+        self,
+        circuit: cirq.AbstractCircuit,
+        problem_topo: NamedTopology,
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py]: done

File: cirq-google/cirq_google/workflow/qubit_placement.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            circuit: The circuit.
+            problem_topo: The topologies (i.e. connectivity) of the circuit.
+            shared_rt_info: A `cg.SharedRuntimeInfo` object that may contain additional info
+                to inform placement.
+            rs: A `RandomState` to enable pseudo-random placement strategies.
+        """
+
+
+@dataclasses.dataclass(frozen=True)
+class NaiveQubitPlacer(QubitPlacer):
```
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:55]: do you have a suggestion

File: cirq-google/cirq_google/workflow/quantum_runtime.py (hunk 5 above)
Comment (tanujkhattar) [reply cirq-google/cirq_google/workflow/quantum_runtime.py:251]: Do we need this annotation? Wouldn't the type of `exe` get inferred automatically based on the type of `executable_group` which we iterating on?

File: cirq-core/cirq/devices/named_topologies.py [hunk 10]
```diff
@@ -20,6 +20,10 @@
 import networkx as nx
 from matplotlib import pyplot as plt
 
+from cirq import _compat
+from cirq.devices import GridQubit, LineQubit
```
Comment (tanujkhattar) [inline cirq-core/cirq/devices/named_topologies.py]: Remove Duplicate imports? See lines 23-24 below

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tanujkhattar) [reply cirq-google/cirq_google/workflow/qubit_placement.py:55]: Maybe `NoOpQubitPlacer`?

**Review by MichaelBroughton — APPROVED** [review]
LGTM once the duplicate import is gone.

File: cirq-google/cirq_google/workflow/qubit_placement.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/qubit_placement.py:55]: it's not really a "no op", it's more like an identity

File: cirq-google/cirq_google/workflow/quantum_runtime.py (hunk 5 above)
Comment (mpharrigan) [reply cirq-google/cirq_google/workflow/quantum_runtime.py:251]: pycharm wasn't picking it up

File: cirq-core/cirq/devices/named_topologies.py (hunk 10 above)
Comment (mpharrigan) [reply cirq-core/cirq/devices/named_topologies.py]: see #4701 but I don't think there's any post merge duplication

**Pull request merged**